xbmc/addons/interfaces/GUI/DialogFileBrowser.cpp

```diff
@@ -0,0 +1,338 @@
+  if (additionalShare)
+    GetVECShares(vecShares, additionalShare, strPath);
+  bool bRet = CGUIDialogFileBrowser::ShowAndGetSource(strPath, allowNetworkShares, &vecShares, strType);
+  if (bRet)
+    strncpy(path, strPath.c_str(), maxStringSize);
+  return bRet;
+}
+
+bool Interface_GUIDialogFileBrowser::show_and_get_image(void* kodiBase, const char* shares, const char* heading,
+                                                        char* path, unsigned int maxStringSize)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid data", __FUNCTION__);
+    return false;
+  }
+
+  if (!shares || !heading)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid handler data (shares='%p', heading='%p') on addon '%s'", __FUNCTION__, shares, heading, addon->ID().c_str());
+    return false;
+  }
+
+  std::string strPath = path;
+
+  VECSOURCES vecShares;
+  GetVECShares(vecShares, shares, strPath);
+  bool bRet = CGUIDialogFileBrowser::ShowAndGetImage(vecShares, heading, strPath);
+  if (bRet)
+    strncpy(path, strPath.c_str(), maxStringSize);
+  return bRet;
+}
+
+bool Interface_GUIDialogFileBrowser::show_and_get_image_list(void* kodiBase, const char* shares, const char* heading, char*** path, unsigned int* entries)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid data", __FUNCTION__);
+    return false;
+  }
+
+  if (!shares || !heading)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid handler data (shares='%p', heading='%p') on addon '%s'", __FUNCTION__, shares, heading, addon->ID().c_str());
+    return false;
+  }
+
+  VECSOURCES vecShares;
+  GetVECShares(vecShares, shares, "");
+
+  std::vector<std::string> pathsInt;
+  bool bRet = CGUIDialogFileBrowser::ShowAndGetImageList(vecShares, heading, pathsInt);
+  if (bRet)
+  {
+    *entries = pathsInt.size();
+    *path = static_cast<char**>(malloc(*entries*sizeof(char*)));
+    for (unsigned int i = 0; i < *entries; ++i)
+      (*path)[i] = strdup(pathsInt[i].c_str());
+  }
+  else
+    *entries = 0;
+  return bRet;
+}
+
+void Interface_GUIDialogFileBrowser::clear_list(void* kodiBase, char*** path, unsigned int entries)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  if (!addon)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid data", __FUNCTION__);
+    return;
+  }
+
+  if (*path)
+  {
+    for (unsigned int i = 0; i < entries; ++i)
+    {
+      if ((*path)[i])
+        free((*path)[i]);
+    }
+    free(*path);
+  }
+  else
+    CLog::Log(LOGERROR, "Interface_GUIDialogFileBrowser::%s - invalid handler data (path='%p') on addon '%s'", __FUNCTION__, path, addon->ID().c_str());
+}
+
+void Interface_GUIDialogFileBrowser::GetVECShares(VECSOURCES& vecShares, std::string strShares, std::string strPath)
```

Comment:
<code>const std::string& strShares</code>, <code>const std::string& strPath</code> ?! No reason for pass-by-value, imo.